python/ray/train/predictor.py

```diff
@@ -139,28 +138,47 @@ def predict(self, data: DataBatchType, **kwargs) -> DataBatchType:
         Args:
             data: A batch of input data of type ``DataBatchType``.
             kwargs: Arguments specific to predictor implementations. These are passed
-            directly to ``_predict_pandas``.
+            directly to ``_predict_numpy`` or ``_predict_pandas``.
 
         Returns:
             DataBatchType:
                 Prediction result. The return type will be the same as the input type.
         """
-        data_df = convert_batch_type_to_pandas(data, self._cast_tensor_columns)
-
         if not hasattr(self, "_preprocessor"):
             raise NotImplementedError(
                 "Subclasses of Predictor must call Predictor.__init__(preprocessor)."
             )
-
         if self._preprocessor:
-            data_df = self._preprocessor.transform_batch(data_df)
+            data = self._preprocessor.transform_batch(data)
+        try:
+            batch_format = TYPE_TO_ENUM[type(data)]
+        except KeyError:
+            raise RuntimeError(
+                f"Invalid input data type of {type(data)}, supported "
+                f"types: {list(TYPE_TO_ENUM.keys())}"
+            )
 
-        predictions_df = self._predict_pandas(data_df, **kwargs)
-        return convert_pandas_to_batch_type(
-            predictions_df,
-            type=TYPE_TO_ENUM[type(data)],
-            cast_tensor_columns=self._cast_tensor_columns,
-        )
+        has_predict_numpy = self.__class__._predict_numpy != Predictor._predict_numpy
+
+        if batch_format == BatchFormat.PANDAS:
+            predictions = self._predict_pandas(data, **kwargs)
+            return convert_pandas_to_batch_type(
+                predictions,
+                BatchFormat.PANDAS,
+                cast_tensor_columns=self._cast_tensor_columns,
+            )
+        elif batch_format == BatchFormat.NUMPY and has_predict_numpy:
+            return self._predict_numpy(data, **kwargs)
+        else:
+            # Fallback to convert to pandas batch and call _predict_pandas
+            # ex: xgboost predict with np.ndarray batch data
```

Comment:
I don't think we should have a fallback path...we should just fail if we get one of the unsupported types.

Reply:
we have a number of existing examples / tests that will fail on this case, unless we decide to re-write them all in the same PR ..